### PR TITLE
feat(vps.snapshot): add download vps snapshot feature

### PR DESCRIPTION
**Pull request by brofik**
| Question         | Answer
| ---------------- | ---
| Branch?          | `release/sprint-2248-2250` 
| Bug fix?         | no
| New feature?     | yes
| Breaking change? | no
| Tickets          | Fix #MANAGER-8766 
| License          | BSD 3-Clause

- [x] Try to keep pull requests small so they can be easily reviewed.
- [x] Commits are signed-off
- [x] Only FR translations have been updated
- [x] Branch is up-to-date with target branch
- [x] Lint has passed locally
- [x] Standalone app was ran and tested locally
- [x] Ticket reference is mentioned in linked commits (internal only)
- [ ] Breaking change is mentioned in relevant commits

### CODE REVIEW

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/index.scss [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    min-height: 56px;
+    width: 100%;
+    margin-top: 16px;
+    margin-bottom: 16px;
+  }
+
+  .button-container {
+    display: flex;
+    justify-content: center;
+  }
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/index.scss:12]: **nitpick:** Avoid writing custom css. You can use utilities available from Bootstrap.

**Review by anooparveti — CHANGES_REQUESTED** [review]

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/index.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import '@uirouter/angularjs';
+
+import routing from './snapshot-download.routing';
+import component from './snapshot-download.component';
+
+const moduleName = 'snapshotDownloadModule';
+angular
+  .module(moduleName, ['ui.router'])
+  .config(routing)
+  .component('snapshotDownload', component)
```
[the remark this answers is not in the document]
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/index.js]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/index.scss (hunk 1 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/index.scss]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  trackClick(hit) {
+    this.atInternet.trackClick({
+      name: `vps::detail::dashboard::snapshot-download::${hit}`,
+      type: 'navigation',
+    });
+  }
+
+  async generateSnapshotDownloadLink() {
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:32]: **nitpick:** Unnecessary `async`
```suggestion
 generateSnapshotDownloadLink() {
```

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  isDownloadLinkAvailable() {
+    return !!this.downloadLink;
+  }
+
+  static convertSize(bytes, decimals = 2) {
+    if (!+bytes) return '0 Bytes';
+
+    const k = 1024;
+    const dm = decimals < 0 ? 0 : decimals;
+    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:72]: **question:** Does not these need translations?

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    if (this.downloadLink) {
+      const params = new Proxy(new URLSearchParams(this.downloadLink), {
+        get: (searchParams, prop) => searchParams.get(prop),
+      });
+      return moment.unix(params.temp_url_expires).format('LLL');
+    }
+    return null;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:89]: **nitpick:** Dont disable lint rules.
```suggestion

```

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html [hunk 6]
```diff
@@ -0,0 +1,97 @@
+<oui-modal
+    data-on-dismiss="$ctrl.dismissModal()"
+    data-heading="{{ ::'vps_dashboard_snapshot_download_modal_title' | translate }}"
+>
+    <oui-tile
+        data-heading="{{ ::'vps_dashboard_snapshot_download_modal_summary' | translate }}"
+        class="h-100"
+    >
+        <oui-tile-definition data-ng-if="true">
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:9]:
```suggestion
        <oui-tile-definition>
```

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            <oui-tile-term
+                data-translate="vps_dashboard_snapshot_download_modal_service_name"
+            >
+            </oui-tile-term>
+            <oui-tile-description data-ng-bind="::$ctrl.serviceName">
+            </oui-tile-description>
+        </oui-tile-definition>
+
+        <oui-tile-definition data-ng-if="true">
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:18]:
```suggestion
        <oui-tile-definition>
```

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            >
+            </oui-tile-term>
+            <oui-tile-description data-ng-bind="::$ctrl.size">
+            </oui-tile-description>
+        </oui-tile-definition>
+    </oui-tile>
+    <div class="mt-3 mb-3">
+        <div
+            class="notification-box oui-message oui-message_info"
+            ng-if="$ctrl.isLoading"
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:50]:
```suggestion
            data-ng-if="$ctrl.isLoading"
```

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        >
+            <span data-translate="{{$ctrl.getMessage('loading')}}"></span>
+        </div>
+
+        <div
+            class="notification-box oui-message oui-message_success"
+            ng-if="$ctrl.downloadLink"
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:57]:
```suggestion
            data-ng-if="$ctrl.downloadLink"
```

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        >
+            <span
+                data-translate="{{$ctrl.getMessage('success')}}"
+                data-translate-values="{ curlDocumentationLink: $ctrl.curlDocumentationLink }"
+            ></span>
+        </div>
+
+        <div
+            class="notification-box oui-message oui-message_error"
+            ng-if="$ctrl.errorMessage"
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:67]:
```suggestion
            data-ng-if="$ctrl.errorMessage"
```

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        >
+            <span data-translate="{{$ctrl.errorMessage}}"></span>
+        </div>
+    </div>
+
+    <div
+        class="d-flex justify-content-center mt-3"
+        ng-if="!$ctrl.downloadLink"
+        ng-hide="$ctrl.isLoading"
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:76]:
```suggestion
        data-ng-if="!$ctrl.downloadLink"
        data-ng-hide="$ctrl.isLoading"
```

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        <oui-button
+            variant="primary"
+            data-on-click="$ctrl.generateSnapshotDownloadLink()"
+            aria-label="{{ ::'vps_dashboard_snapshot_download_modal_snapshot_button_label' | translate }}"
+            ><span
+                data-translate="vps_dashboard_snapshot_download_modal_snapshot_button_label"
+            ></span>
+        </oui-button>
+    </div>
+    <div ng-if="$ctrl.downloadLink">
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:87]:
```suggestion
    <div data-ng-if="$ctrl.downloadLink">
```

File: packages/manager/modules/vps/src/import/vps.service.js [hunk 13]
```diff
@@ -453,6 +453,14 @@ export default /* @ngInject */ function VpsService(
       .catch((error) => error);
   };
 
+  this.getSnapshotUrl = function getSnapshotUrl(serviceName) {
+    return $http
+      .get(`/vps/${serviceName}/snapshot/download`)
+      .then(({ data }) => {
+        return data;
+      });
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/import/vps.service.js:461]:
```suggestion
      .then(({ data }) => data);
```

**Review by anooparveti — CHANGES_REQUESTED** [review]

File: packages/manager/modules/vps/src/dashboard/vps-dashboard.controller.js [hunk 14]
```diff
@@ -34,6 +35,7 @@ export default class {
   ) {
     this.$filter = $filter;
     this.$q = $q;
+    this.$http = $http;
```
Comment (ghyenne) [inline packages/manager/modules/vps/src/dashboard/vps-dashboard.controller.js:38]:
```suggestion

```
It doesn't use.

File: packages/manager/modules/vps/src/import/vps.service.js [hunk 15]
```diff
@@ -453,6 +453,14 @@ export default /* @ngInject */ function VpsService(
...
+  };
+
```
Comment (ghyenne) [inline packages/manager/modules/vps/src/import/vps.service.js:463]:
```suggestion
  this.getSnapshotUrl = function getSnapshotUrl(serviceName) {
    return $http
      .get(`/vps/${serviceName}/snapshot/download`)
      .then(({ data }) => data);
  };

```

File: packages/manager/modules/vps/src/vps/translations/Messages_fr_FR.json [hunk 16]
```diff
@@ -99,6 +99,16 @@
...
   "vps_dashboard_monitoring_sla_ips_tooltip": "Pour le bon fonctionnement de nos outils de monitoring, merci d'autoriser ces plages d'IPs dans votre firewall.",
+  "vps_dashboard_snapshot_download_modal_title": "Download Snapshot",
+  "vps_dashboard_snapshot_download_modal_summary": "Informations",
+  "vps_dashboard_snapshot_download_modal_service_name": "Nom du service",
+  "vps_dashboard_snapshot_download_modal_snapshot_size": "Taille",
+  "vps_dashboard_snapshot_download_modal_snapshot_creation_date": "Date de création du snapshot",
+  "vps_dashboard_snapshot_download_modal_snapshot_link_expiration_date": "Date d'expiration du lien de téléchargement",
+  "vps_dashboard_snapshot_download_modal_snapshot_loading_message": "Génération de la commande de téléchargement du snapshot ...",
+  "vps_dashboard_snapshot_download_modal_snapshot_curl_success": "Commande de téléchargement générée avec succes.",
+  "vps_dashboard_snapshot_download_modal_snapshot_curl_info": "Veilliez copier la command CURL ci-dessus et l'exécuter sur votre terminal. Pour en savoir plus sur CURL <a href=\"{{curlDocumentationLink}}\" target=\"_blank\">Cliquez ici</a>.",
```
Comment (ghyenne) [inline packages/manager/modules/vps/src/vps/translations/Messages_fr_FR.json:110]:
```suggestion
  "vps_dashboard_snapshot_download_modal_snapshot_curl_info": "Veuillez copier la commande CURL ci-dessus et l'exécuter sur votre terminal. Pour en savoir plus sur CURL <a href=\"{{curlDocumentationLink}}\" target=\"_blank\">Cliquez ici</a>.",
```
and you can remove / move the html code

**Review by ghyenne — CHANGES_REQUESTED** [review]

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js (hunk 3 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:32]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html (hunk 6 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:9]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html (hunk 7 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:18]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html (hunk 9 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:57]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html (hunk 10 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:67]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html (hunk 11 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:76]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html (hunk 12 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:87]: Done.

File: packages/manager/modules/vps/src/import/vps.service.js (hunk 13 above)
Comment (brofik) [reply packages/manager/modules/vps/src/import/vps.service.js:461]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:32]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+  getMessage(status) {
+    switch (status) {
+      case 'loading':
```
[the remark this answers is not in the document]
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:92]: Done

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    this.creationDate = moment(this.tabSummary.snapshot.creationDate).format(
+      'LLL',
+    );
+    this.linkExpirationDate = this.getLinkExpirationDate();
+    this.downloadLink = null;
+    this.downloadLinkCurlCommand = null;
+    this.isLoading = false;
+    this.notificationMessage = null;
+    this.errorMessage = null;
+    this.curlDocumentationLink = 'https://curl.se/docs/manual.html';
```
[the remark this answers is not in the document]
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:22]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+        break;
+      default:
+        this.notificationMessage = null;
+    }
+    return this.notificationMessage;
+  }
+
+  dismissModal() {
+    return this.goBack();
+  }
```
[the remark this answers is not in the document]
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:108]: Tracking will be added to this section.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:76]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html (hunk 12 above)
[the remark this answers is not in the document]
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:87]: Done.

File: packages/manager/modules/vps/src/dashboard/vps-dashboard.controller.js (hunk 14 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/vps-dashboard.controller.js:38]: Done.

File: packages/manager/modules/vps/src/import/vps.service.js (hunk 15 above)
Comment (brofik) [reply packages/manager/modules/vps/src/import/vps.service.js:463]: Done.

File: packages/manager/modules/vps/src/vps/translations/Messages_fr_FR.json (hunk 16 above)
Comment (brofik) [reply packages/manager/modules/vps/src/vps/translations/Messages_fr_FR.json:110]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js (hunk 4 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:72]: Not in the current specification.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+            this.errorMessage = `Error ${error.status}:  ${error.statusText}`;
+            break;
+          case 503:
+            this.errorMessage = `Error ${error.status}:  ${error.statusText}`;
+            break;
+          default:
+            this.errorMessage = `Error ${error.status}:  ${error.message}`;
+        }
+      });
+  }
```
[the remark this answers is not in the document]
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js:62]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js (hunk 5 above)
Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.controller.js]: Done.

File: packages/manager/modules/vps/src/dashboard/vps-dashboard.controller.js [hunk 21]
```diff
@@ -246,6 +246,21 @@ export default class {
...
       },
+      download: {
+        text: this.$translate.instant(
+          'vps_configuration_download_snapshot_title_button',
+        ),
+        callback: () => {
+          return this.$state.go('vps.detail.snapshot.download', {
+            serviceName: this.serviceName,
+          });
+        },
```
Comment (ghyenne) [inline packages/manager/modules/vps/src/dashboard/vps-dashboard.controller.js:257]:
```suggestion
        callback: () => this.$state.go('vps.detail.snapshot.download', {
            serviceName: this.serviceName,
          }),
```

Comment (brofik) [reply packages/manager/modules/vps/src/dashboard/vps-dashboard.controller.js]: Done.

File: packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+        <oui-tile-definition data-ng-if="$ctrl.size">
+            <oui-tile-term
+                data-translate="vps_dashboard_snapshot_download_modal_snapshot_size"
+            >
+            </oui-tile-term>
+            <oui-tile-description data-ng-bind="::$ctrl.size">
+            </oui-tile-description>
+        </oui-tile-definition>
+    </oui-tile>
+    <div class="mt-3 mb-3">
```
Comment (anooparveti) [inline packages/manager/modules/vps/src/dashboard/modal/snapshot-download/snapshot-download.html:47]:
```suggestion
    <div class="my-3">
```

**Review by anooparveti — APPROVED** [review]

**Review by ghyenne — APPROVED** [review]

**Pull request closed without merging**